### PR TITLE
feat(sdk): website

### PR DESCRIPTION
**Pull request by tsuf239**
creating a static website resource! 

fixes https://github.com/winglang/wing/issues/1294
fixes https://github.com/winglang/wing/issues/1293

The website resource is implemented by a bucket connected to CloudFront distribution for aws, and an express server for the sim. 
It serves a static folder, and supports adding dynamic content via the `add_json` method.

demo:
https://www.loom.com/share/30a8cf02694440549a0d1d1fd6a0c7d2

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: apps/wing/src/commands/compile.test.ts
Comment (staycoolcall911) [inline apps/wing/src/commands/compile.test.ts:1]: Looks like you have a formatting issue in your local env. All these spaces before/after import names should remain

File: examples/tests/valid/website_with_api.w
Comment (staycoolcall911) [inline examples/tests/valid/website_with_api.w:37]: Another nit: in REST successful post returns 201, not 200
```suggestion
      status: 201
```

File: examples/tests/valid/website_with_api.w
Comment (staycoolcall911) [inline examples/tests/valid/website_with_api.w:31]: nit: http code for bad user input is 400 (bad request). 
In general 400-s are for user error, 500-s are for server error
```suggestion
            status: 400
```

File: examples/tests/valid/website_with_api.w
Comment (staycoolcall911) [inline examples/tests/valid/website_with_api.w:49]: nit: I believe it should be a 204
```suggestion
        status: 204
```

**Review by staycoolcall911 — COMMENTED** [review]
Very cool resource @tsuf239 - I love it that it already supports CORS and has a CDN. I added a few comments, but this needs more review from others on our team

File: libs/wingsdk/test/target-tf-aws/__snapshots__/website.test.ts.snap
Comment (tsuf239) [reply libs/wingsdk/test/target-tf-aws/__snapshots__/website.test.ts.snap:401]: it's the same with the bucket: https://github.com/winglang/wing/blob/b8168d8486ce0948cd9c3c04240702414bb2f207/libs/wingsdk/test/target-tf-aws/__snapshots__/bucket.test.ts.snap#L94

File: tools/hangar/__snapshots__/test_corpus/redis.w_test_sim.md
Comment (MarkMcCulloh) [reply tools/hangar/__snapshots__/test_corpus/redis.w_test_sim.md:1]: Redis tests are actually skipped in ci, but I am surprised it skipped and deleted it locally. I would say undo this change for now

File: libs/wingsdk/src/cloud/website.ts
Comment (MarkMcCulloh) [inline libs/wingsdk/src/cloud/website.ts:45]: This should probably throw instead of "", if this is meant to be the correct path

File: libs/wingsdk/src/cloud/website.ts
Comment (MarkMcCulloh) [inline libs/wingsdk/src/cloud/website.ts:50]: Based on the constructor, this is absolute

```suggestion
   * Absolute local path to the website's static files.
```

File: libs/wingsdk/test/target-tf-aws/__snapshots__/website.test.ts.snap
Comment (tsuf239) [reply libs/wingsdk/test/target-tf-aws/__snapshots__/website.test.ts.snap]: I change the names anyway now, to be other than their namespaces, in my last commit

File: tools/hangar/__snapshots__/test_corpus/website.w_compile_tf-aws.md
Comment (Chriscbr) [reply tools/hangar/__snapshots__/test_corpus/website.w_compile_tf-aws.md:37]: If you're referring to the two "Default" names, this is a side effect of how the SDK layers resources together internally right now to make `cloud.App` behave the same for all targets. For the `tf-aws` target, "root" corresponds to `cdktf.TerraformStack`, one of the "Default" components is corresponds to `CdktfApp` (which we cheated to make it a child of TerraformStack), and the other "Default" component corresponds to the `$Root` construct generated in preflight.js. There might be some ways to remove either/both "Default" layers, but I don't think it's a critical issue

File: apps/wing/src/commands/compile.test.ts
Comment (Chriscbr) [reply apps/wing/src/commands/compile.test.ts:18]: Makes sense to me 👍  I know when folks are using the Wing CLI, we have to be careful about removing files as to not delete people's terraform state files, but I don't think we need to provide the same protection for unit tests so this should be fine

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (Chriscbr) [inline libs/wingsdk/src/target-tf-aws/website.ts:30]:
```suggestion
    new S3BucketWebsiteConfiguration(this, "BucketWebsiteConfiguration", {
```

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (Chriscbr) [inline libs/wingsdk/src/target-tf-aws/website.ts:38]:
```suggestion
    const distribution = new CloudfrontDistribution(this, "Distribution", {
```

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (Chriscbr) [inline libs/wingsdk/src/target-tf-aws/website.ts:83]: I don't think including the name of the terraform resource is needed / usually that info is redundant
```suggestion
    new S3Object(this, `File-${filePath}`, {
```

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (Chriscbr) [inline libs/wingsdk/src/target-tf-aws/website.ts:95]: I'm not sure if we can only use the basename, because then:
* /foo/hello.txt
* /bar/hello.txt

would end up with the same construct ID. Maybe we can use the full file path, or add a hash of the full path, e.g.
```suggestion
    new S3Object(this, `File-${basename}-${hash(filePath).slice(-8)}`, {
```

File: libs/wingsdk/src/cloud/website.ts
Comment (Chriscbr) [inline libs/wingsdk/src/cloud/website.ts:69]: Let's avoid the hairiness of case conversion by naming the parameter `path` for now
```suggestion
  /**
   * Add a JSON file with custom values during the website's deployment.
   * If the path conflicts with file path from the website's static assets, an error will be thrown.
   * @param path the file path it will be uploaded as
   * @param data the data to write to the file
   */
  abstract addJson(path: string, data: Json): string;
```

**Review by Chriscbr — COMMENTED** [review]
Looks great so far! 😄

File: libs/wingsdk/src/cloud/website.ts
Comment (hasanaburayyan) [inline libs/wingsdk/src/cloud/website.ts:51]: The example is a relative path, should that be updated or is relative acceptable and the comment be updated.

File: libs/wingsdk/src/target-sim/website.inflight.ts
Comment (hasanaburayyan) [reply libs/wingsdk/src/target-sim/website.inflight.ts:44]: (thinking outloud) I think staying away from a fixed port is probably better for now, otherwise we might need to consider how to increment ports based on number of sites, and if port is in use do we fail or keep incrementing ports to find a valid `next` port. Also we may end up in some race conditions if we start to get opinionated on ports. 

Thought I do agree the port number changing on sim restarts could be frustrating.

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (hasanaburayyan) [reply libs/wingsdk/src/target-tf-aws/website.ts:28]: nit: I think Chris meant PascalCase rather than cameCase :smile: and I agree thats probably a good style choice to enforce

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (hasanaburayyan) [inline libs/wingsdk/src/target-tf-aws/website.ts:40]: hmm excuse my typescript confusion its pretty late here :smile: but Im not sure I understand this line. Is it the same as saying 

```js
const additionalCloudFrontProps = this._domain? { aliases: [this._domain] } : {};
const distribution = new CloudfrontDistribution(this, "distribution", {
  enabled: true,
  ...additionalCloudFrontProps
```

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (hasanaburayyan) [inline libs/wingsdk/src/target-tf-aws/website.ts:72]: Do we need a tracking issue?

**Review by hasanaburayyan — COMMENTED** [review]
Super cool stuff 🔥 I added a few comments but did not get to finish, will pick back up in the morning.

File: tools/hangar/__snapshots__/test_corpus/website.w_compile_tf-aws.md
Comment (tsuf239) [reply tools/hangar/__snapshots__/test_corpus/website.w_compile_tf-aws.md]: fixed :)

File: libs/wingsdk/src/target-sim/website.inflight.ts
Comment (tsuf239) [reply libs/wingsdk/src/target-sim/website.inflight.ts:44]: I think the user should choose the port (similar to when they make one from scratch) - do we know what target we are running? 
(so we can do something like `let w = new website({path: "...", domain: target == "sim" ? "localhost:8080" :  "real-domain.com" })`)

File: tools/hangar/__snapshots__/test_corpus/redis.w_test_sim.md
Comment (tsuf239) [reply tools/hangar/__snapshots__/test_corpus/redis.w_test_sim.md:1]: image
seems to be skipped on local too, I'll take a look :)

Comment (tsuf239) [reply tools/hangar/__snapshots__/test_corpus/redis.w_test_sim.md:1]: This code is responsible for that:
https://github.com/winglang/wing/blob/746c0f17a4362b77b9c1095f7f28bc3a8ee238b7/tools/hangar/src/generate_tests.ts#L16-L31
here is the meta-comment:
image
and my machine is the one to skip:
image

+ if a contributor will run it they might have the other skipped platform, and it'll happen again...
What do you think will be the best way to solve it @MarkMcCulloh?

File: libs/wingsdk/src/cloud/website.ts
Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:33]: This property exists only on the tf-aws implementation now, for the sim we just override the URL getter

File: libs/wingsdk/src/cloud/website.ts
Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:57]: good catch!

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (tsuf239) [reply libs/wingsdk/src/target-tf-aws/website.ts:72]: it's actually the default behaviour, I wrote it to myself while coding, removing it then, thanks!

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (tsuf239) [reply libs/wingsdk/src/target-tf-aws/website.ts:40]: exactly!

File: libs/wingsdk/src/cloud/website.ts
Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:46]: Sure!! thanks!

File: examples/tests/valid/website_with_api.w
Comment (tsuf239) [reply examples/tests/valid/website_with_api.w:42]: opened: https://github.com/winglang/wing/issues/2289

File: libs/wingsdk/src/cloud/website.ts
Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:45]: https://github.com/winglang/wing/issues/2290 added

File: apps/wing/src/commands/compile.ts
Comment (tsuf239) [reply apps/wing/src/commands/compile.ts]: I agree :) 
should I run the formatted on this (never formatted apparently) folder and create another PR after this one will be merged?

File: libs/wingsdk/src/cloud/website.ts
Comment (staycoolcall911) [inline libs/wingsdk/src/cloud/website.ts:1]: Noticed another issue when working with the website resource on aws:
Need to invalidate cache on CloudFront after every change to the static website's code.
I think as part of `terraform apply` we should have some logic to invalidate some/all of the static files (need to leave an escape hatch for users to decide to remove specific files/patterns only)

Comment (staycoolcall911) [inline libs/wingsdk/src/cloud/website.ts:1]: Noticed another issue when working with the website resource on aws:
Even after we manually invalidated the CloudFront cache - `terraform apply` didn't publish index.html correctly.

File: tools/hangar/__snapshots__/test_corpus/redis.w_test_sim.md
Comment (MarkMcCulloh) [reply tools/hangar/__snapshots__/test_corpus/redis.w_test_sim.md]: Ah, I think I intended to only skip on CI. I think it would be sufficient to add a check only to respect skipPlatforms when `process.env.CI` is `true`

Comment (tsuf239) [reply tools/hangar/__snapshots__/test_corpus/redis.w_test_sim.md]: perfect! thanks :) 
we already have `process.env.CI` right?

Comment (tsuf239) [reply tools/hangar/__snapshots__/test_corpus/redis.w_test_sim.md]: Done :)

File: libs/wingsdk/src/cloud/website.ts
Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts]: fixed :)

File: libs/wingsdk/src/cloud/website.ts
Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:1]: I think I could solve it when making the policy depends on the bucket

Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:1]: I cannot find a cdktf module to perform cache invalidation :( 
I wonder if I'm missing something
I could find one for cdk:
https://docs.aws.amazon.com/cdk/api/v1/docs/aws-s3-deployment-readme.html#cloudfront-invalidation

Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:1]: I think it takes time and depends on the invalidation method, I'll try implementing somehow the invalidation and we'll see if it persists

Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:1]: and see that it's not trivial in terraform :( 
https://stackoverflow.com/questions/69794727/how-can-i-invalidate-aws-cloudfront-distribution-cache-using-terraform

Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:1]: I think we should open an issue and think what can be the best way to achieve it based on the available tools... @staycoolcall911  WDYT?

Comment (staycoolcall911) [reply libs/wingsdk/src/cloud/website.ts:1]: Yes, we can open a separate issue and leave this to a following PR, but then it must be documented, because every user of the Website resource will run into that sooner or later. Perhaps @Chriscbr, @hasanaburayyan or @eladb will have an idea.

Comment (staycoolcall911) [reply libs/wingsdk/src/cloud/website.ts:1]: Ok, thanks, but please note that we (ShaiA and myself) waited for the cache invalidation to complete, then viewed index.html and it was still outdated.

Comment (Chriscbr) [reply libs/wingsdk/src/cloud/website.ts:1]: Ah right, cache invalidation - the hardest problem in computer science 😅 yes I remember now CDK had some dedicated code for handling this for users... it looks like it was written in Python as part of the BucketDeployment resource [here](https://github.com/aws/aws-cdk/blob/c81d115955dbb27ce873ed7c9d71cc0dc8eacf99/packages/aws-cdk-lib/aws-s3-deployment/lib/lambda/index.py#L130-L131)

I feel like we eventually want to handle this using the `OnDeploy` resource (basically, a `cloud.Function` that runs during deployment to call the AWS SDK and perform a cache invalidation with minimal permissions). But the approach mentioned on StackOverflow (creating a local_exec resource) sounds okay as a workaround - documenting as @staycoolcall911 mentioned also works 👍

Comment (tsuf239) [reply libs/wingsdk/src/cloud/website.ts:1]: opened an issue: https://github.com/winglang/wing/issues/2404

File: libs/wingsdk/src/target-tf-aws/website.ts
Comment (Chriscbr) [inline libs/wingsdk/src/target-tf-aws/website.ts:82]: nit: what if the user inputs a file path with no extension?

File: libs/wingsdk/test/target-sim/website.test.ts
Comment (Chriscbr) [inline libs/wingsdk/test/target-sim/website.test.ts:78]: nit: Wrap the `expect` only around the addJson() line to make it clear that is the one throwing the error

File: libs/wingsdk/test/util.ts
Comment (Chriscbr) [inline libs/wingsdk/test/util.ts:21]:
```suggestion
export function tfResourcesWithProperty(
```

File: libs/wingsdk/test/util.ts
Comment (Chriscbr) [inline libs/wingsdk/test/util.ts:25]: style: Add return type

File: tools/hangar/__snapshots__/test_corpus/website.w_test_sim.md
Comment (Chriscbr) [inline tools/hangar/__snapshots__/test_corpus/website.w_test_sim.md:3]: Can we also start snapshotting the simulator.json files? I think tracking these would help us detect breaking changes to the console

(maybe this should be a separate issue...)

**Review by Chriscbr — APPROVED** [review]
🎉 🎉 🎉

**Review by staycoolcall911 — APPROVED** [review]
Should be good to go - nice!

File: tools/hangar/__snapshots__/test_corpus/website.w_test_sim.md
Comment (tsuf239) [reply tools/hangar/__snapshots__/test_corpus/website.w_test_sim.md:3]: added: #2427

**Pull request merged**
